onnx/defs/math/defs.cc

```diff
@@ -11,6 +11,68 @@
 
 namespace ONNX_NAMESPACE {
 
+inline int MathOpTwoIntegers(std::string op_type, int a, int b) {
+  if (op_type == "Add") {
+    return a + b;
+  } else if (op_type == "Sub") {
+    return a - b;
+  } else if (op_type == "Mul") {
+    return a * b;
+  }
+  fail_shape_inference("Wrong op_type name for running propagation: ", op_type);
+}
+
+inline void MathOpDataPropagator(DataPropagationContext& ctx, std::string op_type) {
```

Comment:
I think we can handle dim_params slightly better (e.g., if some are known and some are unknown). We should also handle the case where neither dim_param nor dim_value is set. I suggest something like below (also avoiding code duplication):
```cpp
inline void MathOpDataPropagator(DataPropagationContext& ctx, std::string op_type) {
  const auto input_0 = ctx.getInputData(0);
  const auto input_1 = ctx.getInputData(1);
  if (input_0 == nullptr || input_1 == nullptr) {
      return;
  }

  auto size0 = input_0->dim_size();
  auto size1 = input_1->dim_size();

  if ((size0 != size1) && (size0 != 1) && (size1 != 1) {
        fail_shape_inference("Invalid rank for ", op_type, " broadcasting: (",
            size0, ") vs (", size1, ").");
  }

  TensorShapeProto tsp;
  for (int i = 0; i < std::max(size0, size1); ++i) {
    auto& val0 = input_0->dim(size0 == 1 ? 0 : i);
    auto& val1 = input_1->dim(size1 == 1 ? 0 : i);
    auto* result = tsp.mutable_dim()->Add();
    if (val0.has_dim_value() && val1.has_dim_value())
       result->set_dim_value(MathOpTwoIntegers(op_type, val0.dim_value(), val1.dim_value()));
  }

  ctx.addOutputData(0, std::move(tsp));
}
```

Reply:
Thank your for the suggestion! I just finished something like this.